File: cosky-rest-api/src/main/java/me/ahoo/cosky/rest/security/rbac/AuthorizeService.java
/*
 * Copyright [2021-2021] [ahoo wang <ahoowang@qq.com> (https://github.com/Ahoo-Wang)].
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *      http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.ahoo.cosky.rest.security.rbac;

import com.google.common.base.Strings;
import io.jsonwebtoken.ExpiredJwtException;
import lombok.extern.slf4j.Slf4j;
import me.ahoo.cosky.rest.security.CoSkySecurityException;
import me.ahoo.cosky.rest.security.JwtProvider;
import me.ahoo.cosky.rest.security.TokenExpiredException;
import me.ahoo.cosky.rest.security.annotation.AdminResource;
import me.ahoo.cosky.rest.security.annotation.AllowAnonymous;
import me.ahoo.cosky.rest.security.annotation.OwnerResource;
import me.ahoo.cosky.rest.security.user.User;
import me.ahoo.cosky.rest.support.RequestPathPrefix;
import org.springframework.core.annotation.AnnotatedElementUtils;
import org.springframework.http.HttpStatus;
import org.springframework.http.server.reactive.ServerHttpRequest;
import org.springframework.stereotype.Service;
import org.springframework.web.method.HandlerMethod;
import org.springframework.web.reactive.result.method.annotation.RequestMappingHandlerMapping;
import org.springframework.web.server.ServerWebExchange;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.lang.annotation.Annotation;
import java.util.Objects;
import java.util.Optional;

/**
 * @author ahoo wang
 */
@Slf4j
@Service
public class AuthorizeService {
    public static final String CURRENT_USER_KEY = "cosky.currentUser";
    public static final String AUTH_HEADER = "Authorization";
    public static final String QUERY_PARAM_TOKEN = "token";

    private final RBACService rbacService;
    private final JwtProvider jwtProvider;
    private final RequestMappingHandlerMapping requestMappingHandlerMapping;

    public AuthorizeService(RBACService rbacService, JwtProvider jwtProvider, RequestMappingHandlerMapping requestMappingHandlerMapping) {
        this.rbacService = rbacService;
        this.jwtProvider = jwtProvider;
        this.requestMappingHandlerMapping = requestMappingHandlerMapping;
    }

    public static Optional<String> getToken(ServerHttpRequest request) {
        final String headerToken = request.getHeaders().getFirst(AUTH_HEADER);
        if (!Strings.isNullOrEmpty(headerToken)) {
            return Optional.of(headerToken);
        }
        final String queryToken = request.getQueryParams().getFirst(QUERY_PARAM_TOKEN);
        if (!Strings.isNullOrEmpty(queryToken)) {
            return Optional.of(queryToken);
        }
        return Optional.empty();
    }

    public Mono<AuthorizeResult> authorize(ServerWebExchange exchange) throws TokenExpiredException {
        return requestMappingHandlerMapping
                .getHandler(exchange)
                .ofType(HandlerMethod.class)
                .flatMap(handlerMethod -> authorize(exchange, handlerMethod))
                .defaultIfEmpty(AuthorizeResult.UNAUTHORIZED);
    }

    private Mono<AuthorizeResult> authorize(ServerWebExchange exchange, HandlerMethod handlerMethod) {
        ServerHttpRequest request = exchange.getRequest();
        String requestPath = request.getPath().value();

        final Optional<String> accessToken = getToken(request);
        if (hasAnnotation(handlerMethod, AllowAnonymous.class)) {
            return Mono.just(AuthorizeResult.ALLOW_ANONYMOUS);
        }

        if (!accessToken.isPresent()) {
            return Mono.just(AuthorizeResult.UNAUTHORIZED);
        }

        final User user;
        try {
            user = jwtProvider.authorize(accessToken.get());
            exchange.getAttributes().put(CURRENT_USER_KEY, user);
        } catch (ExpiredJwtException expiredJwtException) {
            if (log.isInfoEnabled()) {
                log.info(expiredJwtException.getMessage(), expiredJwtException);
            }
            return Mono.just(AuthorizeResult.UNAUTHORIZED);
        }

        if (hasAnnotation(handlerMethod, AllowAnonymous.class)) {
            return Mono.just(AuthorizeResult.ofAuthorized(user));
        }

        if (user.isAdmin()) {
            return Mono.just(AuthorizeResult.ofAuthorized(user));
        }

        if (hasAnnotation(handlerMethod, OwnerResource.class)) {
            return Mono.just(AuthorizeResult.ofAuthorized(user));
        }

        if (hasAnnotation(handlerMethod, AdminResource.class)) {
            return Mono.just(AuthorizeResult.ofForbidden(user));
        }

        final String namespace = resolveNamespace(requestPath);
        ResourceAction requestAction = new ResourceAction(namespace, Action.ofHttpMethod(Objects.requireNonNull(request.getMethod())));
        return checkRolePermissions(user, requestAction)
                .map(result -> {
                    if (result) {
                        return AuthorizeResult.ofAuthorized(user);
                    }
                    return AuthorizeResult.ofForbidden(user);
                });
    }

    private String resolveNamespace(String requestPath) {
        String namespace = requestPath.substring(RequestPathPrefix.NAMESPACES_PREFIX.length() + 1);
        int splitIdx = namespace.indexOf("/");
        if (splitIdx > 0) {
            return namespace.substring(0, splitIdx);
        }
        return namespace;
    }


    public static Optional<User> getUserOfRequest(ServerWebExchange serverWebExchange) {
        User user = serverWebExchange.getAttribute(CURRENT_USER_KEY);
        return Optional.ofNullable(user);
    }

    public static User getRequiredUserOfRequest(ServerWebExchange serverWebExchange) {
        return getUserOfRequest(serverWebExchange)
                .orElseThrow(() -> new CoSkySecurityException("UNAUTHORIZED"));
    }

    public static boolean hasAnnotation(HandlerMethod handlerMethod, Class<? extends Annotation> annotationType) {
        return AnnotatedElementUtils.hasAnnotation(handlerMethod.getBeanType(), annotationType)
                || handlerMethod.hasMethodAnnotation(annotationType);
    }

    public Mono<Boolean> checkRolePermissions(User user, ResourceAction requestAction) {
        return Flux.fromIterable(user.getRoleBind())
                .flatMap(rbacService::getRole)
                .any(role -> role.check(requestAction));
    }

    public static class AuthorizeResult {

        public static final AuthorizeResult ALLOW_ANONYMOUS = new AuthorizeResult(true, HttpStatus.OK, null);
        public static final AuthorizeResult UNAUTHORIZED = new AuthorizeResult(false, HttpStatus.UNAUTHORIZED, null);
        private final boolean authorized;
        private final HttpStatus status;
        private final User identity;

        public static AuthorizeResult ofAuthorized(User identity) {
            return new AuthorizeResult(true, HttpStatus.OK, identity);
        }

        public static AuthorizeResult ofForbidden(User identity) {
            return new AuthorizeResult(false, HttpStatus.FORBIDDEN, identity);
        }

        public AuthorizeResult(boolean authorized, HttpStatus status, User identity) {
            this.authorized = authorized;
            this.status = status;
            this.identity = identity;
        }

        public boolean isAuthorized() {
            return authorized;
        }

        public HttpStatus getStatus() {
            return status;
        }

        public User getIdentity() {
            return identity;
        }
    }


}

File: cosky-rest-api/src/main/kotlin/me/ahoo/cosky/rest/security/rbac/AuthorizeService.kt
/*
 * Copyright [2021-present] [ahoo wang <ahoowang@qq.com> (https://github.com/Ahoo-Wang)].
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *      http://www.apache.org/licenses/LICENSE-2.0
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package me.ahoo.cosky.rest.security.rbac

import io.jsonwebtoken.ExpiredJwtException
import me.ahoo.cosky.rest.security.JwtProvider
import me.ahoo.cosky.rest.security.TokenExpiredException
import me.ahoo.cosky.rest.security.annotation.AdminResource
import me.ahoo.cosky.rest.security.annotation.AllowAnonymous
import me.ahoo.cosky.rest.security.annotation.OwnerResource
import me.ahoo.cosky.rest.security.rbac.Action.Companion.ofHttpMethod
import me.ahoo.cosky.rest.security.user.User
import me.ahoo.cosky.rest.support.RequestPathPrefix
import org.slf4j.LoggerFactory
import org.springframework.core.annotation.AnnotatedElementUtils
import org.springframework.http.HttpStatus
import org.springframework.http.server.reactive.ServerHttpRequest
import org.springframework.stereotype.Service
import org.springframework.web.method.HandlerMethod
import org.springframework.web.reactive.result.method.annotation.RequestMappingHandlerMapping
import org.springframework.web.server.ServerWebExchange
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import java.util.*

/**
 * Authorize Service.
 *
 * @author ahoo wang
 */
@Service
class AuthorizeService(
    private val rbacService: RbacService,
    private val jwtProvider: JwtProvider,
    private val requestMappingHandlerMapping: RequestMappingHandlerMapping
) {
    @Throws(TokenExpiredException::class)
    fun authorize(exchange: ServerWebExchange): Mono<AuthorizeResult> {
        return requestMappingHandlerMapping
            .getHandler(exchange)
            .ofType(HandlerMethod::class.java)
            .flatMap { authorize(exchange, it) }
            .defaultIfEmpty(AuthorizeResult.UNAUTHORIZED)
    }

    private fun authorize(exchange: ServerWebExchange, handlerMethod: HandlerMethod): Mono<AuthorizeResult> {
        val request = exchange.request
        val accessToken = getToken(request)
        if (hasAnnotation(handlerMethod, AllowAnonymous::class.java)) {
            return Mono.just(AuthorizeResult.ALLOW_ANONYMOUS)
        }
        if (!accessToken.isPresent) {
            return Mono.just(AuthorizeResult.UNAUTHORIZED)
        }
        val user: User
        try {
            user = jwtProvider.authorize(accessToken.get())
            exchange.attributes[CURRENT_USER_KEY] = user
        } catch (expiredJwtException: ExpiredJwtException) {
            if (log.isInfoEnabled) {
                log.info(expiredJwtException.message, expiredJwtException)
            }
            return Mono.just(AuthorizeResult.UNAUTHORIZED)
        }
        if (hasAnnotation(handlerMethod, AllowAnonymous::class.java)) {
            return Mono.just(AuthorizeResult.ofAuthorized(user))
        }
        if (user.isAdmin) {
            return Mono.just(AuthorizeResult.ofAuthorized(user))
        }
        if (hasAnnotation(handlerMethod, OwnerResource::class.java)) {
            return Mono.just(AuthorizeResult.ofAuthorized(user))
        }
        if (hasAnnotation(handlerMethod, AdminResource::class.java)) {
            return Mono.just(AuthorizeResult.ofForbidden(user))
        }
        val requestPath = request.path.value()
        val namespace = resolveNamespace(requestPath)
        val requestAction = ResourceAction(namespace, ofHttpMethod(Objects.requireNonNull(request.method)))
        return checkRolePermissions(user, requestAction)
            .map { result: Boolean ->
                if (result) {
                    return@map AuthorizeResult.ofAuthorized(user)
                }
                AuthorizeResult.ofForbidden(user)
            }
    }

    private fun resolveNamespace(requestPath: String): String {
        val namespace = requestPath.substring(RequestPathPrefix.NAMESPACES_PREFIX.length + 1)
        val splitIdx = namespace.indexOf("/")
        return if (splitIdx > 0) {
            namespace.substring(0, splitIdx)
        } else {
            namespace
        }
    }

    fun checkRolePermissions(user: User, requestAction: ResourceAction): Mono<Boolean> {
        return Flux.fromIterable(user.roleBind)
            .flatMap { roleName: String ->
                rbacService.getRole(
                    roleName
                )
            }
            .any { role: Role ->
                role.check(
                    requestAction
                )
            }
    }

    class AuthorizeResult(val isAuthorized: Boolean, val status: HttpStatus, val identity: User?) {

        companion object {
            val ALLOW_ANONYMOUS = AuthorizeResult(true, HttpStatus.OK, null)
            val UNAUTHORIZED = AuthorizeResult(false, HttpStatus.UNAUTHORIZED, null)
            fun ofAuthorized(identity: User): AuthorizeResult {
                return AuthorizeResult(true, HttpStatus.OK, identity)
            }

            fun ofForbidden(identity: User): AuthorizeResult {
                return AuthorizeResult(false, HttpStatus.FORBIDDEN, identity)
            }
        }
    }

    companion object {
        private val log = LoggerFactory.getLogger(AuthorizeService::class.java)
        const val CURRENT_USER_KEY = "cosky.currentUser"
        const val AUTH_HEADER = "Authorization"
        const val QUERY_PARAM_TOKEN = "token"
        fun getToken(request: ServerHttpRequest): Optional<String> {
            val headerToken = request.headers.getFirst(AUTH_HEADER)

            if (!headerToken.isNullOrEmpty()) {
                return Optional.of(headerToken)
            }
            val queryToken = request.queryParams.getFirst(QUERY_PARAM_TOKEN)
            return if (!queryToken.isNullOrEmpty()) {
                Optional.of(queryToken)
            } else {
                Optional.empty()
            }
        }

        fun getUserOfRequest(serverWebExchange: ServerWebExchange): Optional<User> {
            val user = serverWebExchange.getAttribute<User>(CURRENT_USER_KEY)
            return Optional.ofNullable(user)
        }

        fun getRequiredUserOfRequest(serverWebExchange: ServerWebExchange): User {
            return getUserOfRequest(serverWebExchange)
                .orElseThrow { SecurityException("UNAUTHORIZED") }
        }

        fun hasAnnotation(handlerMethod: HandlerMethod, annotationType: Class<out Annotation>): Boolean {
            return (
                AnnotatedElementUtils.hasAnnotation(handlerMethod.beanType, annotationType) ||
                    handlerMethod.hasMethodAnnotation(annotationType)
                )
        }
    }
}
